libmeepgeom/array-slice-ll.cpp

```diff
@@ -0,0 +1,220 @@
+/***************************************************************/
+/***************************************************************/
+/***************************************************************/
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <complex>
+
+#include "meep.hpp"
+
+#include "ctl-math.h"
+#include "ctlgeom.h"
+
+#include "meepgeom.hpp"
+
+using namespace meep;
+
+typedef std::complex<double> cdouble;
+
+vector3 v3(double x, double y=0.0, double z=0.0)
+{
+  vector3 v;
+  v.x=x; v.y=y; v.z=z;
+  return v;
+}
+
+// passthrough field function
+cdouble default_field_function(const cdouble *fields,
+			const vec &loc, void *data_)
+{
+  (void) loc; // unused
+  (void) data_; // unused
+  return fields[0];
+}
+
+/***************************************************************/
+/***************************************************************/
+/***************************************************************/
+#define RELTOL 1.0e-6
+double Compare(double *d1, double *d2, int N, const char *Name)
+{
+  double MaxRelErr=0.0;
+  for(int n=0; n<N; n++)
+   { 
+     double Scale = fmax( fabs(d1[n]), fabs(d2[n]) );
+     if (Scale<1.0e-8)
+      continue;
+     double RelErr = fabs(d1[n] - d2[n]) / Scale;
```

Comment:
A more reliable way to compare two floating-point arrays is typically to compute ‖d₁-d₂‖ < reltol * (‖d₁‖+‖d₂‖)/2, where ‖⋯‖ denotes the norm (e.g. the Euclidean/L₂ norm) of the whole array.

This avoids problems with scaling and having to arbitrarily drop values below some threshold.   If you want a more strict test that notices errors in individual values, just use the Chebyshev/L_\infty norm for ‖⋯‖.